vault: add known issue for azure self target vuln (#3126)
hashicorp/web-unified-docs: content/vault/global/partials/important-changes/known-issue/azure-static-role-self-target.mdx

| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 2.0.0            | No

<!-- END: Vault:=v2.x -->

<!-- BEGIN: Vault:=v1.21.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.21.0           | No

<!-- END: Vault:=v1.21.x -->

<!-- BEGIN: Vault:=v1.20.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.20.0           | No

<!-- END: Vault:=v1.20.x -->

<!-- BEGIN: Vault:=v1.19.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.19.0           | No

<!-- END: Vault:=v1.19.x -->

#### Issue

The Azure secrets engine does not prevent a role from targeting the same Azure
AD application that Vault uses as its own service principal. Both static roles

> and dynamic roles with an explicit `application_object_id` are affected.

Reviewer: ```suggestion
AD application that Vault uses as its service principal. The problem affects static
and dynamic roles with an explicit `application_object_id
```
Style correction: write in active voice

Clarifying question: is the qualified, "with an explicit ID", meant to apply to both role types? Or are we trying to say it affects static roles and only dynamic roles with an explicit ID?

#### Workaround

**Step 1: Audit for poisoned roles**

Before applying any code fix, audit your deployment for static roles that
already target Vault's own service principal. On Vault Enterprise, repeat
these steps for every namespace because each namespace can have its own

> Azure secrets engine mounts.

Reviewer: ```suggestion
Before applying any code fixes, audit your Azure deployment for static roles that
target the service principal used by Vault.

On Vault Enterprise, repeat the following steps for every namespace with an
Azure secrets engine mount:
```
Style correction: write in active voice, avoid possessives

> First, identify all Azure secrets engine mounts in each namespace:

Reviewer: ```suggestion
1. Identify all  the Azure secrets engine mounts :
```

```shell-session

> # List all secret engine mounts (run once per namespace)

Reviewer: ```suggestion
```
Avoid comments in code examples we want folks to copy/paste

$ vault secrets list -format=json | \
  jq -r 'to_entries[] | select(.value.type == "azure") | .key'
```

For each mount path returned above, audit both static and dynamic roles:

```shell-session
# List static roles for a given mount
$ vault list <mount>/static-roles

# For each static role, inspect its application_object_id
$ vault read <mount>/static-roles/<name>

# List dynamic roles for a given mount
$ vault list <mount>/roles

# For each dynamic role, inspect its application_object_id
$ vault read <mount>/roles/<name>

# Compare against the engine's configured client_id
$ vault read <mount>/config

> ```

Reviewer: ````suggestion

1. Fetch and save all the Azure mount paths under the current namespace:

    ```shell-session
    export MOUNT_PATHS=$(                                          \
    vault secrets list -format=json |                            \
    jq -r 'to_entries[] | select(.value.type == "azure") | .key' \
    )
    ```

1. Compare the client ID (`client_id`) for each Azure mount path to the
   application object ID (`application_object_id`) for the available static
   roles:

    ```shell-session
    $ for mount in ${MOUNT_PATHS} ; do 
        CLIENTID=$(vault read -format=json ${mount}config | jq '.data.client_id')
        echo "${mount} ($CLIENTID) static roles:"
        vault list -format=json ${mount}static-roles | \
        jq --arg client ${CLIENTID} -r 'to_entries[] | select(.value.application_object_id == "$client") | .key'
    done
    ```

1. Compare the client ID (`client_id`) for each Azure mount path to the
   application object ID (`application_object_id`) for the available dynamic
   roles :

    ```shell-session
    $ for mount in ${MOUNT_PATHS} ; do 
        CLIENTID=$(vault read -format=json ${mount}config | jq '.data.client_id')
        echo "${mount} ($CLIENTID) dynamic roles:"
        vault list -format=json ${mount}roles | \
        jq --arg client ${CLIENTID} -r 'to_entries[] | select(.value.application_object_id == "$client") | .key'
    done
    ```
````

Any static role whose `application_object_id` resolves to the engine's own
Azure app registration must be deleted immediately:

```shell-session
$ vault delete <mount>/static-roles/<name>

> ```

Reviewer: ````suggestion
1. Delete any static role whose `application_object_id` matches the client ID
   (Azure app ID) of the associated mount path:

    ```shell-session
    $ vault delete <mount>/static-roles/<name>
    ```
````

After deleting a static role, revoke the corresponding credentials directly in
Azure AD by removing the associated application password (client secret) from

> the app registration. Do not rely on Vault lease revocation for this step.

Reviewer: ```suggestion
1. Revoke the corresponding credentials directly in Azure AD by removing the
   associated application password (client secret) from the app registration.
   You cannot rely on Vault lease revocation to remove access for the static role.
```

Any dynamic role whose `application_object_id` resolves to the engine's own
Azure app registration must also be deleted:

```shell-session
$ vault delete <mount>/roles/<name>

> ```

Reviewer: ````suggestion
1. Delete any dynamic role whose `application_object_id` matches the client ID
   (Azure app ID) of the associated mount path:

    ```shell-session
    $ vault delete <mount>/roles/<name>
    ```
````

For dynamic roles, also revoke all active leases so that any already-issued
credentials are removed from the Azure app registration immediately rather than
waiting for lease expiry:

```shell-session
# Revoke all leases issued by the affected dynamic role
$ vault lease revoke -prefix <mount>/creds/<name>

> ```

Reviewer: ````suggestion
1. Revoke active leases for the applicable dynamic roles so that Azure removes
   any already-issued credentials from the app registration immediately rather
   than waiting for lease expiry:

    ```shell-session
    $ vault lease revoke -prefix <mount>/creds/<name>
    ```
````

**Step 2: Upgrade**

Upgrade to a patched version of the Azure secrets engine plugin once one is
available. The patch rejects static and dynamic role creation and updates that
target Vault's own application, and blocks credential reads and rotations for

> any pre-existing poisoned roles.

Reviewer: ```suggestion
Upgrade to a patched version of the Azure secrets engine plugin once we release
it. The patch rejects static and dynamic role creation and updates that target
the application used by Vault, and blocks credential reads and rotations for
any pre-existing poisoned roles.
```
Style correction: write in active voice

<!-- BEGIN: Vault:=v1.20.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.20.0           | No

<!-- END: Vault:=v1.20.x -->

<!-- BEGIN: Vault:=v1.19.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.19.0           | No

<!-- END: Vault:=v1.19.x -->

#### Issue

The Azure secrets engine does not prevent a role from targeting the same Azure
AD application that Vault uses as its service principal. The problem affects static
and dynamic roles with an explicit `application_object_id

#### Workaround

**Step 1: Audit for poisoned roles**

Before applying any code fixes, audit your Azure deployment for static roles that
target the service principal used by Vault.

On Vault Enterprise, repeat the following steps for every namespace with an
Azure secrets engine mount:

1. Fetch and save all the Azure mount paths under the current namespace:

    ```shell-session

>     export MOUNT_PATHS=$(                                          \

Reviewer: ```suggestion
    $ export MOUNT_PATHS=$(                                          \
```

| Known issue | Open     | Enterprise    | 1.21.0           | No

<!-- END: Vault:=v1.21.x -->

<!-- BEGIN: Vault:=v1.20.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.20.0           | No

<!-- END: Vault:=v1.20.x -->

<!-- BEGIN: Vault:=v1.19.x -->

| Change      | Status   | Vault edition | Affected version | Fixed version
| ----------- | -------- | ------------- | ---------------- | -------------
| Known issue | Open     | Enterprise    | 1.19.0           | No

<!-- END: Vault:=v1.19.x -->

#### Issue

The Azure secrets engine does not prevent a role from targeting the same Azure
AD application that Vault uses as its service principal. The problem affects static
and dynamic roles with an explicit `application_object_id

#### Workaround

**Step 1: Audit for poisoned roles**

Before applying any code fixes, audit your Azure deployment for static roles that
target the service principal used by Vault.

On Vault Enterprise, repeat the following steps for every namespace with an
Azure secrets engine mount:

> 1. Fetch and save all the Azure mount paths under the current namespace:

Reviewer: ````suggestion
1. Fetch and save all the Azure mount paths under the current namespace:

    ```shell-session
    export MOUNT_PATHS=$(                                          \
    vault secrets list -format=json |                            \
    jq -r 'to_entries[] | select(.value.type == "azure") | .key' \
    )
    ```

1. Compare the client ID (`client_id`) for each Azure mount path to the
   application object ID (`application_object_id`) for the available static
   roles:

    ```shell-session
    $ for mount in ${MOUNT_PATHS} ; do 
        CLIENTID=$(vault read -format=json ${mount}config | jq '.data.client_id')
        echo "${mount} ($CLIENTID) static roles:"
        vault list -format=json ${mount}static-roles | \
        jq --arg client ${CLIENTID} -r 'to_entries[] | select(.value.application_object_id == "$client") | .key'
    done
    ```

1. Compare the client ID (`client_id`) for each Azure mount path to the
   application object ID (`application_object_id`) for the available dynamic
   roles :

    ```shell-session
    $ for mount in ${MOUNT_PATHS} ; do 
        CLIENTID=$(vault read -format=json ${mount}config | jq '.data.client_id')
        echo "${mount} ($CLIENTID) dynamic roles:"
        vault list -format=json ${mount}roles | \
        jq --arg client ${CLIENTID} -r 'to_entries[] | select(.value.application_object_id == "$client") | .key'
    done
    ```
````

> 1. Delete any static role whose `application_object_id` matches the client ID
>    (Azure app ID) of the associated mount path:

Reviewer: ````suggestion
1. Delete any static role whose `application_object_id` matches the client ID
   (Azure app ID) of the associated mount path:

    ```shell-session
    $ vault delete <mount>/static-roles/<name>
    ```
````

    

1. Revoke the corresponding credentials directly in Azure AD by removing the
   associated application password (client secret) from the app registration.
   You cannot rely on Vault lease revocation to remove access for the static role.

1. Delete any dynamic role whose `application_object_id` matches the client ID

>    (Azure app ID) of the associated mount path:

Reviewer: ````suggestion
1. Delete any dynamic role whose `application_object_id` matches the client ID
   (Azure app ID) of the associated mount path:

    ```shell-session
    $ vault delete <mount>/roles/<name>
    ```
````

> 1. Revoke active leases for the applicable dynamic roles so that Azure removes
>    any already-issued credentials from the app registration immediately rather
>    than waiting for lease expiry:

Reviewer: ````suggestion
1. Revoke active leases for the applicable dynamic roles so that Azure removes
   any already-issued credentials from the app registration immediately rather
   than waiting for lease expiry:

    ```shell-session
    $ vault lease revoke -prefix <mount>/creds/<name>
    ```
````
